optuna/samplers/_nsga2/crossover.py

```diff
@@ -0,0 +1,503 @@
+                raise RuntimeError(
+                    "undxm must be n > m-2,"
+                    " where m is the number of parental individuals to be selected"
+                    " at crossover (m=4) and n is the number of parameters of the individuals."
+                )
+
+            sigma_xi = 1 / np.sqrt(_m)
+            sigma_eta = (
+                0.35
+                * np.sqrt(_m + 1)
+                * np.sqrt(3)
+                / np.sqrt(_n - _m)
+                / np.sqrt(_m + 2)
+                / np.sqrt(2)
+            )
+            params_array = _undxm(parents_numerical_params_array, rng, sigma_xi, sigma_eta)
+        elif crossover_name == "spx":
+            epsilon = np.sqrt(len(parents_numerical_params_array[0]) + 2)
+            params_array = _spx(parents_numerical_params_array, rng, epsilon)
+        else:
+            assert False
+
+        params = numerical_transform.untransform(params_array)
+        for param_name in params.keys():
+            child[param_name] = params[param_name]
+
+        if _is_contained(child, search_space):
+            break
+
+    return child
+
+
+def select_m(crossover_name: str) -> int:
+    # Select the number of parent individuals to be used for crossover.
+    if crossover_name in ["uniform", "blxalpha", "sbx", "vsbx"]:
+        m = 2
+    elif crossover_name in ["undx", "spx"]:
+        m = 3
+    elif crossover_name in ["undxm"]:
+        m = 4
+    else:
+        assert False
+    return m
+
+
+def _select_parents(
+    crossover_name: str,
+    study: Study,
+    parent_population: Sequence[FrozenTrial],
+    rng: np.random.RandomState,
+    dominates: Callable[[FrozenTrial, FrozenTrial, Sequence[StudyDirection]], bool],
+) -> List[FrozenTrial]:
+    m = select_m(crossover_name)
+    parents = []
+    for _ in range(m):
+        parent = _select_parent(
+            study, [t for t in parent_population if t not in parents], rng, dominates
+        )
+        parents.append(parent)
+    return parents
+
+
+def _select_parent(
+    study: Study,
+    parent_population: Sequence[FrozenTrial],
+    rng: np.random.RandomState,
+    dominates: Callable[[FrozenTrial, FrozenTrial, Sequence[StudyDirection]], bool],
+) -> FrozenTrial:
+    # TODO(ohta): Consider to allow users to specify the number of parent candidates.
+    population_size = len(parent_population)
+    candidate0 = parent_population[rng.choice(population_size)]
+    candidate1 = parent_population[rng.choice(population_size)]
+
+    # TODO(ohta): Consider crowding distance.
+    if dominates(candidate0, candidate1, study.directions):
+        return candidate0
+    else:
+        return candidate1
+
+
+def _uniform(
+    x1: np.ndarray, x2: np.ndarray, rng: np.random.RandomState, swapping_prob: float
+) -> np.ndarray:
+    # https://www.researchgate.net/publication/201976488_Uniform_Crossover_in_Genetic_Algorithms
+    # Section 1 Introduction
+
+    assert x1.shape == x2.shape
+    assert len(x1.shape) == 1
```

Comment:
`ndim` might be convenient.
```suggestion
    assert x1.ndim == 1
```

Reply:
Ditto for similar other occurrences.